internal/controller/postgrescluster/volumes.go

```diff
@@ -122,6 +131,555 @@ func (r *Reconciler) observePersistentVolumeClaims(
+// in the given volume to the expected location before the PostgresCluster is
+// bootstrapped. It returns any errors and a boolean indicating whether the
+// main control loop should continue or return early to allow time for the job
+// to complete.
+func (r *Reconciler) reconcileMovePGDataDir(ctx context.Context,
+	cluster *v1beta1.PostgresCluster, moveJobs *batchv1.JobList) (bool, error) {
+
+	moveDirJob := &batchv1.Job{}
+	moveDirJob.ObjectMeta = naming.MovePGDataDirJob(cluster)
+
+	// check for an existing Job
+	for i := range moveJobs.Items {
+		if moveJobs.Items[i].Name == moveDirJob.Name {
+			if jobCompleted(&moveJobs.Items[i]) {
+				// if the Job is completed, return as this only needs to run once
+				return false, nil
+			}
+			if !jobFailed(&moveJobs.Items[i]) {
+				// if the Job otherwise exists and has not failed, return and
+				// give the Job time to finish
+				return true, nil
+			}
+		}
+	}
+
+	// at this point, the Job either wasn't found or it has failed, so the it
+	// should be created
+	moveDirJob.ObjectMeta.Annotations = naming.Merge(cluster.Spec.Metadata.
+		GetAnnotationsOrNil())
+	labels := naming.Merge(cluster.Spec.Metadata.GetLabelsOrNil(),
+		naming.DirectoryMoveJobLabels(cluster.Name),
+		map[string]string{
+			naming.LabelMovePGDataDir: "",
+		})
+	moveDirJob.ObjectMeta.Labels = labels
+
+	script := fmt.Sprintf(`echo "Preparing cluster %s volumes for PGO v5.x"
+    echo "pgdata_pvc=%s"
+    echo "Current PG data directory volume contents:" 
+    ls -lh "/pgdata"
+    echo "Now updating PG data directory..."
+    [ -d "/pgdata/%s" ] && mv "/pgdata/%s" "/pgdata/pg%s_bootstrap"
+    rm -f "/pgdata/pg%s/patroni.dynamic.json"
+    echo "Updated PG data directory contents:" 
+    ls -lh "/pgdata"
+    echo "PG Data directory preparation complete"
+    `, cluster.Name,
+		cluster.Spec.DataSource.ExistingVolumes.ExistingPGDataVolume.PVCName,
+		cluster.Spec.DataSource.ExistingVolumes.ExistingPGDataVolume.Directory,
+		cluster.Spec.DataSource.ExistingVolumes.ExistingPGDataVolume.Directory,
+		strconv.Itoa(cluster.Spec.PostgresVersion),
+		strconv.Itoa(cluster.Spec.PostgresVersion))
+
+	jobSpec := &batchv1.JobSpec{
+		Template: v1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{Labels: labels},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{{
+					Command:      []string{"bash", "-ceu", script},
+					Image:        config.PostgresContainerImage(cluster),
+					Name:         naming.ContainerJobMovePGDataDir,
+					VolumeMounts: []corev1.VolumeMount{postgres.DataVolumeMount()},
+				}},
+				SecurityContext: postgres.PodSecurityContext(cluster),
+				// Set RestartPolicy to "Never" since we want a new Pod to be
+				// created by the Job controller when there is a failure
+				// (instead of the container simply restarting).
+				RestartPolicy:                v1.RestartPolicyNever,
+				AutomountServiceAccountToken: initialize.Bool(false),
```

Comment:
Since these Jobs don't make Kubernetes API calls, we can just use the default ServiceAccount and mount the credentials. We do the same thing for [pgBouncer](https://github.com/CrunchyData/postgres-operator/blob/master/internal/controller/postgrescluster/pgbouncer.go#L442-L444).

Reply:
ah, i see. Do you mind adding that comment here too :bow:

Reply:
:+1: sure thing